Change default for decoy library generation

hroest opened the pull request:
Currently our decoy library generation is quite poor since we only have two options

- shuffle the whole protein
- reverse the whole protein

However, this strategy does not conserve the overall mass distribution of peptides and may introduce biases into the target-decoy approach. It seems better to be able to shuffle and/or reverse on peptide level with the possibility of keeping either C or N terminus constant. This PR will introduce this as default option. The new options will enable users to 

- shuffle with either C, N, CN or no terminus constant
- reverse with either C, N, CN or no terminus constant

One downside of this is that very short peptides (<3-4 AA) cannot be properly reversed, e.g. `PEK` reversed with C/N terminus constant will still be `PEK` while a peptide like `PETK` will simply become `PTEK` which is highly similar.

Another important point for redundant databases is whether the decoy and target DB have the same number of unique peptides. If not, there will be biases towards the decoy database as there are many more unique peptides than in the target DB (since we shuffle with different seeds each time). This PR ensures that the same seed is used for each peptide, resolving this problem. Alternatively, we could also keep a list of all peptides and their mapping to a decoy peptide to solve this problem, but this may be quite memory intensive.

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp:
```diff
@@ -43,60 +43,68 @@
-    pattern.push_back("K");
-    pattern.push_back("R");
-    pattern.push_back("P");
+    defaults_.setValue("non_shuffle_pattern", "K,R,P", "Residues to not shuffle (keep at a constant position when shuffling). Separate by comma, e.g. use 'K,P,R' here.");
```
maybe directly use a stringlist?

timosachsenberg commented on src/utils/DecoyDatabase.cpp:
```diff
@@ -153,33 +224,61 @@ class TOPPDecoyDatabase :
+          {
+            String temp;
+            Size x = protein.sequence.size();
+            srand(time(nullptr));
```
why don't we use seed here?

timosachsenberg commented:
Hi,
Nice PR. 
Also worth to consider would be a hybrid approach in the future: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4941923/

timosachsenberg commented:
+1 from my side if seed is added. @cbielow any comments?

cbielow commented on src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp:
```diff
@@ -43,60 +43,68 @@
-      }
-    }
-    return idx;
+    defaults_.setValue("keepN", "true", "Whether to keep N terminus constant", ListUtils::create<String>("advanced"));
```
is this *protein* N/C-term?

cbielow replied:
This means peptide C/N terms, right?! we should make this clear.

cbielow commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMDecoy.h:
```diff
@@ -132,44 +138,70 @@ namespace OpenMS
     */
-    std::vector<std::pair<std::string::size_type, std::string> > findFixedResidues(
-      std::string sequence);
+    IndexType findFixedResidues(std::string sequence);
```
could this be
`IndexType findFixedResidues(const std::string& sequence) const;`?

cbielow commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMDecoy.h:
```diff
@@ -132,44 +138,70 @@ namespace OpenMS
     */
-    std::vector<std::pair<std::string::size_type, std::string> > findFixedAndTermResidues(
-      std::string sequence);
+    IndexType findFixedAndTermResidues(std::string sequence) const;
```
could this be
`IndexType findFixedAndTermResidues(const std::string& sequence) const;`?

cbielow commented on src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp:
```diff
@@ -43,60 +43,68 @@
     return idx;
   }
 
+  MRMDecoy::IndexType MRMDecoy::findFixedResidues(std::string sequence)
```
can also be const?

cbielow commented on src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp:
```diff
@@ -258,61 +265,70 @@ namespace OpenMS
     }
+  }
+
+  OpenMS::TargetedExperiment::Peptide MRMDecoy::pseudoreversePeptide(
```
is `pseudo` a common term in this context?
If not, maybe there is something more descriptive? (reversePeptideConstCTerm()?)

hroest replied:
we have been using this in the targeted proteomics world quite a bit, so I opt to leave it.

hroest commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMDecoy.h:
```diff
@@ -132,44 +138,70 @@ namespace OpenMS
+                const OpenMS::TargetedExperiment::Peptide& peptide,
+                const bool keepN,
+                const bool keepC, 
+                const std::vector<String> & const_pattern = std::vector<String>());
```
no real reason to use a vector or string (except that it was there before) but it does make everything easier to write.

hroest commented on src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp:
```diff
@@ -116,8 +124,8 @@ namespace OpenMS
-    OpenMS::TargetedExperiment::Peptide peptide, double identity_threshold, int seed,
-    int max_attempts)
+    OpenMS::TargetedExperiment::Peptide peptide, const double identity_threshold, int seed,
+    const int max_attempts)
```
yes, that could be the next step. Since these methods are more general than just for MRM decoys, that should be the next step. However, we have to think about whether we then still want to use `OpenMS::TargetedExperiment::Peptide` or switch to `AASequence`

hroest commented:
I think I addressed all the comments. One question remains: is it possible to let the user provide a user-defined regex or does the user have to modify Enzymes.xml for this? I dont think I got it working as most methods supporting this are private / protected. That seems rather user-unfriendly?

timosachsenberg commented:
I think it is intended that the user edits the Enzymes.xml. That way a workflow can trace an Enzyme from start to end using our internal file formats. I agree that is is not super user-friendly but a way to ensure the evidence trace for enzymes. This also makes EnzymesDB read-only and we don't run into threading issues as we do with the residue db.

cbielow commented on src/tests/class_tests/openms/source/MRMDecoy_test.cpp:
```diff
@@ -377,15 +439,14 @@ START_SECTION(OpenMS::TargetedExperiment::Peptide pseudoreversePeptide(OpenMS::T
-
 END_SECTION
 
 START_SECTION(OpenMS::TargetedExperiment::Peptide reversePeptide(OpenMS::TargetedExperiment::Peptide peptide))
```
technically, these are `[EXTRA]` now (since protected), but I'm not sure what the status of the checker is anyway.
Maybe discuss at retreat (will add it to the list).

cbielow commented:
only StopWatch test failed.
+1 for merge
